Sources/Build/BuildPlan.swift

```diff
@@ -1680,76 +1700,154 @@ public class BuildPlan {
-    ) throws -> [(product: ResolvedProduct, targetBuildDescription: SwiftTargetBuildDescription)] {
-        guard case .manifest(let generate) = buildParameters.testDiscoveryStrategy else {
+    ) throws -> [(product: ResolvedProduct, discoveryTargetBuildDescription: SwiftTargetBuildDescription?, manifestTargetBuildDescription: SwiftTargetBuildDescription)] {
+        guard case .manifest(let enableDiscovery, let isCustomManifest) = buildParameters.testDiscoveryStrategy else {
             throw InternalError("makeTestManifestTargets should not be used for build plan with useTestManifest set to false")
         }
 
-        var generateRedundant = generate
-        var result: [(ResolvedProduct, SwiftTargetBuildDescription)] = []
+        var isDiscoveryEnabledRedundantly = enableDiscovery && !isCustomManifest
+        var result: [(ResolvedProduct, SwiftTargetBuildDescription?, SwiftTargetBuildDescription)] = []
         for testProduct in graph.allProducts where testProduct.type == .test {
             guard let package = graph.package(for: testProduct) else {
                 throw InternalError("package not found for \(testProduct)")
             }
-            generateRedundant = generateRedundant && nil == testProduct.testManifestTarget
-            // if test manifest exists, prefer that over test detection,
+            isDiscoveryEnabledRedundantly = isDiscoveryEnabledRedundantly && nil == testProduct.testManifestTarget
+            // If a non-custom test manifest exists, prefer that over test detection,
             // this is designed as an escape hatch when test discovery is not appropriate
             // and for backwards compatibility for projects that have existing test manifests (LinuxMain.swift)
             let toolsVersion = graph.package(for: testProduct)?.manifest.toolsVersion ?? .v5_5
-            if let testManifestTarget = testProduct.testManifestTarget, !generate {
-                let desc = try SwiftTargetBuildDescription(
+
+            let defaultTestManifestFilename = SwiftTarget.testManifestNames.first!
+            if testProduct.testManifestTarget != nil && enableDiscovery && !isCustomManifest {
+                let testManifestFileName = testProduct.underlyingProduct.testManifest?.basename ?? defaultTestManifestFilename
+                observabilityScope.emit(warning: "'--enable-test-discovery' was specified so the '\(testManifestFileName)' manifest file for '\(testProduct.name)' will be ignored and a manifest will be generated automatically. To use test discovery with a custom manifest, pass '--experimental-test-manifest-path <file>'.")
+            } else if testProduct.testManifestTarget == nil, isCustomManifest, let customTestManifestPath = testProduct.underlyingProduct.testManifest {
+                observabilityScope.emit(error: "'--experimental-test-manifest-path' was specified but the file '\(customTestManifestPath)' could not be found.")
+            }
```

Comment:
probably worth explaining in code comment the idea and role of the (newly introduced) customTestManifestPath